### PR TITLE
SNOW-876389: compatibility with bcr-1057 on outer lateral table join

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-876389, fix compatibility with [BCR-1057](https://docs.snowflake.com/en/release-notes/bcr-bundles/2023_04/bcr-1057)

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

[BCR-1057]: https://snowflakecomputing.atlassian.net/browse/BCR-1057?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/snowflake/sqlalchemy/base.py [hunk 1]
```diff
@@ -88,6 +100,331 @@
...
+                    ),
+                )
+
+    @sa_util.preload_module("sqlalchemy.sql.util")
+    def _join_determine_implicit_left_side(self, raw_columns, left, right, onclause):
+        """When join conditions don't express the left side explicitly,
+        determine if an existing FROM or entity in this query
+        can serve as the left hand side.
+
+        """
```
Comment (sfc-gh-mkeller) [inline src/snowflake/sqlalchemy/base.py:162]:
```suggestion
        """Determine whether an existing FROM or entity in this query can serve as the LHS.

        This is useful when join conditions don't express the left side explicitly.
        """
```

File: src/snowflake/sqlalchemy/base.py [hunk 2]
```diff
@@ -88,6 +100,331 @@
...
+            if len(indexes) == 1:
+                replace_from_obj_index = indexes[0]
+                left = from_clauses[replace_from_obj_index]
+        else:
+            potential = {}
+            statement = self.statement
+
+            for from_clause in itertools.chain(
+                itertools.chain.from_iterable(
+                    [element._from_objects for element in raw_columns]
```
Comment (sfc-gh-mkeller) [inline src/snowflake/sqlalchemy/base.py:181]: You could also use `map(operators.itemgetter('_from_objects', raw_colums)`

File: src/snowflake/sqlalchemy/base.py [hunk 3]
```diff
@@ -88,6 +100,331 @@
...
+                ),
+                itertools.chain.from_iterable(
+                    [element._from_objects for element in statement._where_criteria]
```
Comment (sfc-gh-mkeller) [inline src/snowflake/sqlalchemy/base.py:184]: Same here

File: src/snowflake/sqlalchemy/base.py [hunk 4]
```diff
@@ -88,6 +100,331 @@
...
+@sql.base.CompileState.plugin_for("orm", "select")
+class SnowflakeORMSelectCompileState(context.ORMSelectCompileState):
+    def _join_determine_implicit_left_side(
+        self, entities_collection, left, right, onclause
+    ):
+        """When join conditions don't express the left side explicitly,
+        determine if an existing FROM or entity in this query
+        can serve as the left hand side.
+
+        """
```
Comment (sfc-gh-mkeller) [inline src/snowflake/sqlalchemy/base.py:227]:
```suggestion
        """Determine if an existing FROM or entity in this query can serve as LHS.
        
        This is useful when join conditions don't express the left side explicitly.
        """
```

File: src/snowflake/sqlalchemy/base.py [hunk 5]
```diff
@@ -88,6 +100,331 @@
...
+        create_aliases,
+        aliased_generation,
+        outerjoin,
+        full,
+    ):
+        """given raw "left", "right", "onclause" parameters consumed from
+        a particular key within _join(), add a real ORMJoin object to
+        our _from_obj list (or augment an existing one)
+
+        """
```
Comment (sfc-gh-mkeller) [inline src/snowflake/sqlalchemy/base.py:344]:
```suggestion
        """Add a real ORMJoin object to our _from_obj list, or augment an existing one.
        
        Given raw "left", "right", "onclause" parameters consumed from
        a particular key within _join().
        """
```

File: tests/test_orm.py [hunk 6]
```diff
@@ -326,3 +337,66 @@ class User(Base):
                 assert user.fullname == "test_user"
             finally:
                 Base.metadata.drop_all(con)
+
+
+def test_outer_lateral_join(engine_testaccount):
+    pytest.skip("test case needs to be fixed")
```
Comment (sfc-gh-mkeller) [inline tests/test_orm.py:343]:
```suggestion
```

File: tests/test_orm.py [hunk 7]
```diff
@@ -326,3 +337,66 @@ class User(Base):
...
+    e2 = Employee(employee_id=102, last_name="Paulson", department_id=1)
+    e3 = Employee(employee_id=103, last_name="Johnson", department_id=2)
+    d1 = Department(department_id=1, name="Engineering")
+    d2 = Department(department_id=2, name="Support")
+    session.add_all([e1, e2, e3, d1, d2])
+    session.commit()
+
+    sub = select(Department).lateral()
+    query = select(Employee.employee_id).select_from(Employee).outerjoin(sub)
+    session.execute(query)
```
Comment (sfc-gh-mkeller) [inline tests/test_orm.py:371]: Assert the results

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

File: src/snowflake/sqlalchemy/base.py [hunk 8]
```diff
@@ -88,6 +100,331 @@
...
+- _join_left_to_right
+"""
+
+
+# handle Snowflake BCR bcr-1057
+@CompileState.plugin_for("default", "select")
+class SnowflakeSelectState(SelectState):
+    def _setup_joins(self, args, raw_columns):
+        for (right, onclause, left, flags) in args:
+            isouter = flags["isouter"]
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/sqlalchemy/base.py:116]: this is the original sqlalchemy source code, so I prefer keeping this, sqlalchemy might has some code handling it

File: src/snowflake/sqlalchemy/base.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/sqlalchemy/base.py:181]: this is copying sqlalchemy source code, so I'd rather keep it

File: src/snowflake/sqlalchemy/base.py (hunk 4 above)
Comment (sfc-gh-aling) [reply src/snowflake/sqlalchemy/base.py:227]: copied pasted from sqlalchemy: https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/selectable.py#L4848-L4850

File: src/snowflake/sqlalchemy/base.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/sqlalchemy/base.py:344]: copied pasted from sqlalchemy: https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/context.py#L1854

File: src/snowflake/sqlalchemy/base.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/sqlalchemy/base.py:162]: copied pasted from sqlalchemy: https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/selectable.py#L4848-L4850

File: tests/test_orm.py (hunk 7 above)
Comment (sfc-gh-aling) [reply tests/test_orm.py]: updated

File: tests/test_orm.py (hunk 6 above)
Comment (sfc-gh-aling) [reply tests/test_orm.py]: updated

**Pull request merged**